File: 분할 정복/11444/11444.py
def dot(a, b):
   x, y, z = len(a), len(b[0]), len(b)
   result = []
   for i in range(x):
      temp = []
      for j in range(y):
         m = 0
         for k in range(z):
            m += a[i][k]*b[k][j]
         temp.append(m % 1000000007)
      result.append(temp)
   return result

n = int(input())
arr = [[0, 1], [1, 1]]
answer = [[1, 0], [0, 1]]
while n > 0:
   if n % 2:
      answer = dot(answer, arr)
   arr = dot(arr, arr)
   n //= 2
print(answer[0][1])

File: 분할 정복/11444/11444.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<vector<long>> dot(vector<vector<long>> &, vector<vector<long>> &);

int main()
{
   long long n;
   cin >> n;
   vector<vector<long>> v = {{0, 1}, {1, 1}};
   vector<vector<long>> answer = {{1, 0}, {0, 1}};
   
   while (n)
   {
      if (n % 2)
         answer = dot(answer, v);
      v = dot(v, v);
      n /= 2;
   }
   cout << answer[0][1];

   return 0;
}

vector<vector<long>> dot(vector<vector<long>> &a, vector<vector<long>> &b)
{
   int x = a.size(), y = b[0].size(), z = b.size();
   vector<vector<long>> result(x, vector<long>(y, 0));
   for (int i = 0; i < x; ++i)
      for (int j = 0; j < y; ++j)
      {
         for (int k = 0; k < z; ++k)
            result[i][j] += a[i][k] * b[k][j];
         result[i][j] %= 1000000007;
      }
   return result;
}